File: c/search/string/Sunday.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int *getNext(char *text, int length, int nextSize, int minValue);

int sunday(char *text, char *pattern) {
    int length = strlen(text);
    int patternLength = strlen(pattern);

    int maxValue = text[0];
    int minValue = text[0];
    for (char *p = text + 1; '\0' != *p; ++p) {
        if (maxValue < *p) {
            maxValue = *p;
        }
        if (minValue > *p) {
            minValue = *p;
        }
    }
    int *next = getNext(pattern, patternLength, maxValue - minValue + 1, minValue);

    int index = 0;
    while (index <= length - patternLength) {
        int patternIndex = 0;
        while (text[index + patternIndex] == pattern[patternIndex]) {
            ++patternIndex;
            if (patternIndex == patternLength) {
                return index;
            }
        }

        if (index == length - patternLength) {
            return -1;
        }
        index += next[text[index + patternLength] - minValue];
    }
    return -1;
}

static int *getNext(char *text, int length, int nextSize, int minValue) {
    int *next = (int *) malloc(sizeof(int) * nextSize);
    for (int i = 0; i < nextSize; ++i) {
        next[i] = length + 1;
    }
    for (int i = 0; i < length; ++i) {
        int index = text[i] - minValue;
        if (index >= 0 && index < nextSize) {
            next[index] = length - i;
        }
    }
    return next;
}

File: java/search/string/Sunday.java
package search.string;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Sunday {
    public int search(String text, String pattern) {
        List<Character> aux = Arrays.asList(text.chars().mapToObj(c -> (char) c).toArray(Character[]::new));
        int maxValue = Collections.max(aux);
        int minValue = Collections.min(aux);
        int[] next = getNext(pattern, maxValue - minValue + 1, minValue);

        int index = 0;
        while (index <= text.length() - pattern.length()) {
            int patternIndex = 0;
            while (text.charAt(index + patternIndex) == pattern.charAt(patternIndex)) {
                ++patternIndex;
                if (patternIndex == pattern.length()) {
                    return index;
                }
            }

            if (index == text.length() - pattern.length()) {
                return -1;
            }
            index += next[text.charAt(index + pattern.length()) - minValue];
        }
        return -1;
    }

    private int[] getNext(String text, int nextSize, int minValue) {
        int[] next = new int[nextSize];
        Arrays.fill(next, text.length() + 1);
        for (int i = 0; i < text.length(); ++i) {
            int index = text.charAt(i) - minValue;
            if (index >= 0 && index < nextSize) {
                next[index] = text.length() - i;
            }
        }
        return next;
    }
}
